### PR TITLE
Adds the evolution tower

### PR DESCRIPTION
**Pull request by TiviPlus**
## About The Pull Request

See changelog

## Why It's Good For The Game

comtent

## Changelog
:cl:
add: Added the evolution tower xeno building, which costs 300 points and when built generates a extra .25 evo points per evo tick
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 1]
```diff
@@ -17,6 +17,7 @@
 	var/list/dead_xenos = list() // xenos that are still assigned to this hive but are dead.
 	var/list/ssd_xenos
 	var/list/list/xenos_by_zlevel = list()
+	var/list/obj/structure/xeno/evotower/evotowers = list()
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:20]: autodoc, also could be a int if you're not doing a range check

File: code/modules/xenomorph/xeno_structures.dm [hunk 2]
```diff
@@ -1050,3 +1050,39 @@ TUNNEL
 	light_initial_color = LIGHT_COLOR_PURPLE
 	ammo = /datum/ammo/xeno/sticky
 	firerate = 5
+
+
+/obj/structure/xeno/evotower
```
Comment (BraveMole) [inline code/modules/xenomorph/xeno_structures.dm:1055]: a range check with other evo towers maybe? Also, marines should have a way of finding those, wether with the pool tracker, or it emitting sound, or it emitting light

File: code/modules/xenomorph/xeno_structures.dm [hunk 3]
```diff
@@ -1050,3 +1050,39 @@ TUNNEL
...
+	obj_integrity = 600
+	max_integrity = 600
+	///hivenumber of this tower
+	var/hivenumber
+	///boost amt to be added per tower per cycle
+	var/boost_amount = 0.25
+
+/obj/structure/xeno/evotower/Initialize(mapload, hivenum)
+	. = ..()
+	GLOB.hive_datums[hivenum].evotowers += src
```
Comment (BraveMole) [inline code/modules/xenomorph/xeno_structures.dm:1071]: if you're not doing a range check, you don't need to store ref in a glob list, a counter is enough

File: code/modules/xenomorph/xeno_structures.dm [hunk 4]
```diff
@@ -1050,3 +1050,39 @@ TUNNEL
...
+	hivenumber = hivenum
+
+/obj/structure/xeno/evotower/Destroy()
+	GLOB.hive_datums[hivenumber].evotowers -= src
+	return ..()
+
+/obj/structure/xeno/xeno_turret/ex_act(severity)
```
Comment (BraveMole) [inline code/modules/xenomorph/xeno_structures.dm:1078]: wrong path

File: code/modules/xenomorph/xeno_structures.dm [hunk 5]
```diff
@@ -1050,3 +1050,39 @@ TUNNEL
...
+	switch(severity)
+		if(EXPLODE_DEVASTATE)
+			take_damage(700)
+		if(EXPLODE_HEAVY)
+			take_damage(500)
+		if(EXPLODE_LIGHT)
+			take_damage(300)
+
+/obj/structure/xeno/evotower/get_xeno_hivenumber()
```
Comment (BraveMole) [inline code/modules/xenomorph/xeno_structures.dm:1087]: should be with other get_xeno_hivenumber proc

File: code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm [hunk 6]
```diff
@@ -95,6 +95,47 @@ GLOBAL_LIST_INIT(upgrade_categories, list("Buildings", "Defences", "Xenos"))//,
...
+/datum/hive_upgrade/building/evotower/on_buy(mob/living/carbon/xenomorph/buyer)
+	if(!do_after(buyer, 10 SECONDS, TRUE, buyer, BUSY_ICON_BUILD))
+		return
+
+	if(!can_buy(buyer, FALSE))
+		return
+
+	var/turf/buildloc = get_step(buyer, SOUTHWEST)
+
+	if(SSpoints.xeno_points_by_hive[buyer.hivenumber] < psypoint_cost)
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm:128]: this is in the can buy IIRC

File: code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm [hunk 7]
```diff
@@ -95,6 +95,47 @@ GLOBAL_LIST_INIT(upgrade_categories, list("Buildings", "Defences", "Xenos"))//,
...
+	var/turf/buildloc = get_step(buyer, SOUTHWEST)
+	if(!buildloc)
+		return FALSE
+
+	if(buildloc.density)
+		if(!silent)
+			to_chat(buyer, span_xenowarning("You cannot build in a dense location!"))
+		return FALSE
+
+/datum/hive_upgrade/building/evotower/on_buy(mob/living/carbon/xenomorph/buyer)
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm:119]: honestly all these on_buy procs could be one proc on the parent proc
you just need a time to build, a building type, you already have the psy cost

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 8]
```diff
@@ -200,6 +201,12 @@
 		xenos += ssd_xeno
 	return xenos
 
+///fetches number of bonus points given to the hive,
+/datum/hive_status/proc/get_evolution_boost()
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:205]: just get the lenght, this proc is uneeded

**Review by BraveMole — CHANGES_REQUESTED** [review]
Not a huge fan of the concept, i'd rather it being a local effect maybe? 

But we'll see, that's good enough for now.

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm (hunk 1 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/hive_datum.dm:20]: done

File: code/modules/xenomorph/xeno_structures.dm (hunk 2 above)
Comment (TiviPlus) [reply code/modules/xenomorph/xeno_structures.dm:1055]: done

File: code/modules/xenomorph/xeno_structures.dm (hunk 3 above)
Comment (TiviPlus) [reply code/modules/xenomorph/xeno_structures.dm:1071]: I want to add potentially tier two of these or such

File: code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm (hunk 7 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm:119]: They have different build requirements

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm (hunk 8 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/hive_datum.dm:205]: I want to add potentially tier two of these or such

File: code/modules/xenomorph/xeno_structures.dm (hunk 5 above)
Comment (TiviPlus) [reply code/modules/xenomorph/xeno_structures.dm:1087]: done

File: code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm (hunk 6 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm:128]: done

File: code/modules/xenomorph/xeno_structures.dm (hunk 4 above)
Comment (TiviPlus) [reply code/modules/xenomorph/xeno_structures.dm:1078]: done

**Comment by David-Stormwell** [comment]
Good stuff.
👍

File: code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm (hunk 7 above)
Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm:119]: not really

image

image

The range check can be done on the child proc, but all the rest is identical

Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm:119]: also, are you sure you want to allow xeno to put up all towers close to each other?

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm (hunk 7 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm:119]: image

**Comment by GradusNL** [comment]
> 
> 
> Not a huge fan of the concept, i'd rather it being a local effect maybe?
> 
> But we'll see, that's good enough for now.

Why would a local effect be good? You'd either have to dramatically lower the cost to the same level as turrets or give them a big enough range that they might as well have a global effect. Placing these on/near the frontline means they are going to get destroyed very easily and not be worth the cost.

File: code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm (hunk 7 above)
Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm:119]: fine by me

**Review by BraveMole — APPROVED** [review]

**Pull request merged**